imap_processing/swe/l1a/swe_science.py

```diff
@@ -0,0 +1,168 @@
+        if key != "SCIENCE_DATA":
+            metadata_arrays.setdefault(key, []).append(value.raw_value)
+
+
+def swe_science(decom_data):
+    """SWE L1A algorithm steps:
+        - Read data from each SWE packet file
+        - Uncompress counts data
+        - Store metadata fields and data in DataArray of xarray
+        - Save data to dataset
+
+    In each packet, SWE collects data for 15 seconds. In each second,
+    it collect data for 12 energy steps and at each energy step,
+    it collects 7 data from each 7 CEMs.
+
+    Each L1A data from each packet will have this shape: 15 rows, 12 columns,
+    and each cell in 15 x 12 table contains 7 element array.
+    These dimension maps to this:
+        15 rows --> 15 seconds
+        12 column --> 12 energy steps in each second
+        7 element --> 7 CEMs counts
+
+    In L1A, we don't do anything besides read raw data, uncompress counts data and
+    store data in 15 x 12 x 7 array.
+
+    SWE want to keep all value as it is in L1A. Post L1A, we group data into full cycle
+    and convert raw data to engineering data as needed.
+
+    Parameters
+    ----------
+    packet_file : str
+        packet file path
+
+    Returns
+    -------
+    xarray.Dataset
+        xarray dataset with data.
+    """
+    science_array = []
+
+    metadata_arrays = collections.defaultdict(list)
+
+    # We know we can only have 8 bit numbers input, so iterate over all
+    # possibilities once up front
+    decompression_table = np.array([uncompress_counts(i) for i in range(256)])
+
+    for data_packet in decom_data:
+        # read raw data
+        binary_data = data_packet.data["SCIENCE_DATA"].raw_value
+        # read binary string to an int and then convert it to
+        # bytes. This is to convert the string to bytes.
+        # Eg. "0000000011110011" --> b'\x00\xf3'
+        byte_data = int(binary_data, 2).to_bytes(1260, byteorder="big")
+        # convert bytes to numpy array of uint8
+        raw_counts = np.frombuffer(byte_data, dtype=np.uint8)
+
+        # Uncompress counts. Uncompressed data is a list of 1260
+        # where 1260 = 15 seconds x 12 energy steps x 7 CEMs
+        # Take the "raw_counts" indices/counts mapping from
+        # decompression_table and then reshape the return
+        uncompress_data = np.take(decompression_table, raw_counts).reshape(15, 12, 7)
+
+        # Save data with its metadata field to attrs and DataArray of xarray.
+        science_array.append(uncompress_data)
+        add_metadata_to_array(data_packet, metadata_arrays)
+
+    science_xarray = xr.DataArray(
+        science_array,
+        dims=["number_of_packets", "seconds", "energy_steps", "cem_counts"],
+    )
+
+    met_time = xr.DataArray(
+        metadata_arrays["SHCOARSE"],
+        name="MET",
+        dims=["number_of_packets"],
+        attrs=dict(
+            description="Mission elapsed time",
+            units="seconds since start of the mission",
+        ),
+    )
```

Comment:
Is this time since launch? At some point we should convert the time to unix timestamp (seconds since 1970).  The xarray_to_cdf routine has a flag to convert unix timestamp to CDF_TT2000 time

Reply:
I totally agree. I am not sure of that myself. I asked Nick Dutton and his guess is that it may be the time when we start getting first series of data packet from spacecraft or time when we start getting science data. We don't know for sure though.

Reply:
That is a good question of when we want to convert time. Because you can't get to unix timestamp from MET without using SPICE. So I think we'll need @laspsandoval's routines to handle that conversion.

Can we convert to Unix time right away in L1a? Or do we need to carry MET around in every file for the SPICE queries? We can carry around both MET and Unix time presumably, just call them different variable names.